les/clientpool.go

```diff
@@ -0,0 +1,592 @@
+func (f *clientPool) saveToDb() {
+	now := f.clock.Now()
+	storage := clientPoolStorage{
+		LogOffset: uint64(f.logOffset(now)),
+	}
+	for _, c := range f.connectedMap {
+		f.finalizeBalance(c, now)
+	}
+	i := 0
+	storage.List = make([]*negBalance, len(f.negBalanceMap))
+	for _, e := range f.negBalanceMap {
+		storage.List[i] = e
+		i++
+	}
+	enc, err := rlp.EncodeToBytes(storage)
+	if err != nil {
+		log.Error("Failed to encode negative balance list", "err", err)
+	} else {
+		f.db.Put([]byte("clientPool"), enc)
+	}
+	for _, b := range f.posBalanceMap {
+		f.storePosBalance(b)
+	}
+}
+
+// storePosBalance stores a single positive balance entry in the database
+func (f *clientPool) storePosBalance(b *posBalance) {
+	if b.value == b.lastStored {
+		return
+	}
+	enc, err := rlp.EncodeToBytes(b)
+	if err != nil {
+		log.Error("Failed to encode client balance", "err", err)
+	} else {
+		f.db.Put(append([]byte("clientPool-balance"), b.id[:]...), enc)
+	}
+}
+
+// getPosBalance retrieves a single positive balance entry from cache or the database
+func (f *clientPool) getPosBalance(id enode.ID) *posBalance {
+	if b, ok := f.posBalanceMap[id]; ok {
+		f.posBalanceQueue.Remove(b.queueIndex)
+		f.posBalanceAccessCounter--
+		f.posBalanceQueue.Push(b, f.posBalanceAccessCounter)
+		return b
+	}
+	balance := &posBalance{}
+	if enc, err := f.db.Get(append([]byte("clientPool-balance"), id[:]...)); err == nil {
+		if err := rlp.DecodeBytes(enc, balance); err != nil {
+			log.Error("Failed to decode client balance", "err", err)
+			balance = &posBalance{}
+		}
+	}
+	balance.id = id
+	balance.queueIndex = -1
+	if f.posBalanceQueue.Size() >= f.queueLimit {
+		b := f.posBalanceQueue.PopItem().(*posBalance)
+		f.storePosBalance(b)
+		delete(f.posBalanceMap, b.id)
+	}
+	f.posBalanceAccessCounter--
+	f.posBalanceQueue.Push(balance, f.posBalanceAccessCounter)
+	f.posBalanceMap[id] = balance
+	return balance
+}
+
+// addBalance updates the positive balance of a client.
+// If setTotal is false then the given amount is added to the balance.
+// If setTotal is true then amount represents the total amount ever added to the
+// given ID and positive balance is increased by (amount-lastTotal) while lastTotal
+// is updated to amount. This method also allows removing positive balance.
+func (f *clientPool) addBalance(id enode.ID, amount uint64, setTotal bool) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	pb := f.getPosBalance(id)
+	c := f.connectedMap[id]
+	var negBalance uint64
+	if c != nil {
+		pb.value, negBalance = c.balanceTracker.getBalance(f.clock.Now())
+	}
+	if setTotal {
+		if pb.value+amount > pb.lastTotal {
+			pb.value += amount - pb.lastTotal
+		} else {
+			pb.value = 0
+		}
+		pb.lastTotal = amount
+	} else {
+		pb.value += amount
+	}
+	if c != nil {
+		c.balanceTracker.setBalance(pb.value, negBalance)
+		if !c.priority && pb.value > 0 {
+			c.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
+		}
```

Comment:
If the final `pb.Value > 0` while original `pb.Value == 0`, we should do more things here:

* If the client is connected emit a `updateCapacity` request to client and also somehow mark it as a priority client;
* If the client is not connected, we shouldn't start the balance tracker.

Reply:
Indeed, I should set the priority flag here (done).
I don't update capacity automatically, it has to be requested externally after the priority status has been granted. The logic is that when you receive priority status your capacity stays at the default value initially. Extra capacity requests will be added later btw, right now everything still works at freeClientCapacity.
Also, if the client is not connected then c is nil so I don't start the balance tracker (it does not exist at all because there is no clientInfo for disconnected clients).